### ISSUE TITLE
Bad state: Invalid DartDevelopmentServiceException JSON error from Linux_android_emu tests causing many flakes

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#152684]
For example from: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android_emu%20external_textures_integration_test/2527/overview

```
[2024-08-01 02:41:47.004421] [STDOUT] stdout: [  +14 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0/, connecting to VM service at http://127.0.0.1:45843/hII3gND_jls=/.
[2024-08-01 02:41:48.230663] [STDOUT] stderr: 
[2024-08-01 02:41:48.230740] [STDOUT] stderr: Bad state: Invalid DartDevelopmentServiceException JSON: {error_code: 2, message: Failed to start Dart Development Service}
[2024-08-01 02:41:48.231206] [STDOUT] stderr: #0      new DartDevelopmentServiceException.fromJson (package:flutter_tools/src/base/dds.dart:144:5)
```

Not sure if it's related, but it looks like this started happening after the API level updated from 34 to 35 in https://github.com/flutter/flutter/commit/c3fbf3dd98f91f46cbc4a5d649209aeaff72576c. cc @camsim99 

Since it seems like the error in parsing the dds json is obscuring the actual failure, also cc @bkonyi and @andrewkolos

**Comment by bkonyi** [comment]
Looks like some bad pattern matching in the DDS exception parsing code. I should have a fix up shortly.

**Comment by bkonyi** [comment]
Fix up for review: https://github.com/flutter/flutter/pull/152685

**Comment by zanderso** [comment]
The parse error is fixed, but tests are still flaking on DDS failing to start:
```
[2024-08-01 13:28:16.838831] [STDOUT] stdout: [  +16 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0/, connecting to VM service at http://127.0.0.1:34965/THrz3rDrL_Q=/.
[2024-08-01 13:28:17.887528] [STDOUT] stdout: [+1048 ms] Warning: Failed to start DDS: Failed to start Dart Development Service
[2024-08-01 13:28:17.889297] [STDOUT] stderr: [        ] DDS has failed to start and there is not an existing DDS instance available to connect to. Please file an issue at https://github.com/flutter/flutter/issues with the following error message:
[2024-08-01 13:28:17.889347] [STDOUT] stderr: 
[2024-08-01 13:28:17.889362] [STDOUT] stderr:             Failed to start Dart Development Service.
[2024-08-01 13:28:17.890347] [STDOUT] stdout: [   +2 ms] "flutter drive" took 117,501ms.
[2024-08-01 13:28:17.894389] [STDOUT] stderr: 
[2024-08-01 13:28:17.894453] [STDOUT] stderr: Bad state: Failed to start Dart Development Service
[2024-08-01 13:28:17.894672] [STDOUT] stderr: #0      DartDevelopmentService.startDartDevelopmentService (package:flutter_tools/src/base/dds.dart:290:9)
[2024-08-01 13:28:17.894707] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.894726] [STDOUT] stderr: #1      FlutterDriverService.reuseApplication (package:flutter_tools/src/drive/drive_service.dart:220:9)
[2024-08-01 13:28:17.894744] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.894761] [STDOUT] stderr: #2      DriveCommand.runCommand (package:flutter_tools/src/commands/drive.dart:279:9)
[2024-08-01 13:28:17.894779] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.894802] [STDOUT] stderr: #3      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1408:27)
[2024-08-01 13:28:17.894820] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.894841] [STDOUT] stderr: #4      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
[2024-08-01 13:28:17.894860] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.894877] [STDOUT] stderr: #5      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
[2024-08-01 13:28:17.894901] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.894919] [STDOUT] stderr: #6      FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:421:9)
[2024-08-01 13:28:17.894942] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.894961] [STDOUT] stderr: #7      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
[2024-08-01 13:28:17.894984] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.895003] [STDOUT] stderr: #8      FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:364:5)
[2024-08-01 13:28:17.895027] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.895047] [STDOUT] stderr: #9      run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:130:9)
[2024-08-01 13:28:17.895069] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.895088] [STDOUT] stderr: #10     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
[2024-08-01 13:28:17.895106] [STDOUT] stderr: <asynchronous suspension>
[2024-08-01 13:28:17.895127] [STDOUT] stderr: #11     main (package:flutter_tools/executable.dart:93:3)
[2024-08-01 13:28:17.895145] [STDOUT] stderr: <asynchronous suspension>
```

**Comment by camsim99** [comment]
I'm not sure what DartDevelopmentService is but I also experienced another socket/connection issue when trying to run tests on Android 35 emulators where `localhost` wouldn't resolves as a host for `Socket`s:

```
java.lang.RuntimeException: java.net.UnknownHostException: Unable to resolve host "localhost": No address associated with hostname
      at dev.flutter.scenariosui.ScreenshotUtil.lambda$onCreate$0(ScreenshotUtil.java:73)
      at dev.flutter.scenariosui.ScreenshotUtil$$ExternalSyntheticLambda1.run(Unknown Source:0)
```

from https://github.com/flutter/engine/blob/30ffe6eb428b76e9a9223d7261ec0e7adece4ad6/testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenariosui/ScreenshotUtil.java#L70

However, `127.0.0.1`. Could that possibly be related?

**Comment by zanderso** [comment]
@bkonyi would you be able to help interpret the error messages above?

**Comment by bkonyi** [comment]
We only throw that exception in two places in DDS:

- [When we encounter a `StateError` when starting the DDS server](https://github.com/dart-lang/sdk/blob/main/pkg/dds/lib/src/dds_impl.dart#L145)
- [If DDS's VM service connection disappears during DDS initialization](https://github.com/dart-lang/sdk/blob/main/pkg/dds/lib/src/dds_impl.dart#L102)

**Comment by bkonyi** [comment]
Looking at the DDS code a bit closer, I don't think DDS failing to bind to an address during initialization would result in the above failure (it should result in a "connection issue" exception in that case). I think the only time we can encounter a state error during initialization is when we have an outstanding JSON-RPC request to the VM service and the connection closes. Are we confident that the target application isn't crashing?

**Comment by zanderso** [comment]
Routing this to the Android team for triage and to try to repro locally.

**Comment by camsim99** [comment]
Just tried running this test with `flutter drive --target lib/frame_rate_main.dart --dds` 20 times locally on an Android 35 emulator and not see this issue (I did see another flake with the test itself, though -- `renders frames from the device at a rate similar to the frames produced`). That emulator I test on is not the same (most likely) than the image that test is using, so it's possible it's a bug with that specific version or otherwise maybe this is impacted by running in CI?

I also recall their being flakes when I first tried bumping the AVD version last year, so I do wonder if this will continue to flake as much (not that this is a good solution).

**Comment by zanderso** [comment]
Yeah, I don't know, but it certainly wouldn't surprise me at all if these flakes were sensitive to both the host platform (i.e. Linux vs. mac) and the exact version of the emulator and Android system image. It's possible that matching those versions exactly would be necessary to get a local repro.

The flake detection bot is very likely to complain about these tests on Wednesday morning, so I'm going to go ahead and bump the priority on this, as well as mark these tests `bringup: true`, which the bot would otherwise be doing then anyway.

**Comment by matanlurey** [comment]
In case it gets lost, `driver_test` (as far as I can tell) should _not_ have been disabled: https://github.com/flutter/flutter/commit/ae66107a5c6fdf92a8d85a45520615a075966600#r145053387.

**Comment by zanderso** [comment]
`driver_test` is also flaking, but possibly not for the same reason. It is currently failing on a late initialization error here:

https://github.com/flutter/flutter/blob/master/dev/integration_tests/android_driver_test/test_driver/main_test.dart#L41

because the tear down code can apparently run even when the set up code doesn't complete successfully (?). I would guess that the late initialization error is masking the real reason for the failure.

However, to answer your other question on the PR, yes, the other tests can/should just disable DDS the same way that `driver_test` does, I think.

**Comment by zanderso** [comment]
I filed a separate issue for `driver_test` here: https://github.com/flutter/flutter/issues/152899

**Comment by zanderso** [comment]
Strangely, after shifting these tests to staging, the flake rate has dropped dramatically. The flake only happened once here: https://ci.chromium.org/ui/p/flutter/builders/staging/Linux_android_emu%20android_defines_test/105/overview. I'm wondering if DDS is somehow not being cleaned up properly.

I'm going to proceed with @matanlurey's suggestion of passing `--no-dds` (which is incidentally also the suggestion made in the help text for the flag...) in https://github.com/flutter/flutter/pull/152898. Once that lands, we can try moving these tests back to prod.

**Comment by zanderso** [comment]
As stated, this issue has been addressed. The root cause was instability in the API Level 35 emulator, but this symptom was addressed by disabling DDS in most tests.
